plugin/trino-hive/src/main/java/io/trino/plugin/hive/CachingDirectoryLister.java

```diff
@@ -82,20 +91,49 @@ private static SchemaTablePrefix parseTableName(String tableName)
     public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)
             throws IOException
     {
-        List<LocatedFileStatus> files = cache.getIfPresent(path);
-        if (files != null) {
-            return simpleRemoteIterator(files);
+        if (!isCacheEnabledFor(table.getSchemaTableName())) {
+            return fs.listLocatedStatus(path);
         }
-        RemoteIterator<LocatedFileStatus> iterator = fs.listLocatedStatus(path);
 
-        if (tablePrefixes.stream().noneMatch(prefix -> prefix.matches(table.getSchemaTableName()))) {
-            return iterator;
+        ValueHolder cachedValueHolder;
+        try {
+            cachedValueHolder = cache.get(path, ValueHolder::new);
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e); // cannot happen
+        }
+        if (cachedValueHolder.getFiles().isPresent()) {
+            return simpleRemoteIterator(cachedValueHolder.getFiles().get());
         }
-        return cachingRemoteIterator(iterator, path);
+        return cachingRemoteIterator(cachedValueHolder, fs, path);
     }
 
-    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(RemoteIterator<LocatedFileStatus> iterator, Path path)
+    @Override
+    public void invalidate(Table table)
     {
+        if (isCacheEnabledFor(table.getSchemaTableName()) && isLocationPresent(table.getStorage())) {
+            if (table.getPartitionColumns().isEmpty()) {
+                cache.invalidate(new Path(table.getStorage().getLocation()));
+            }
+            else {
+                // a partitioned table can have multiple paths in cache
+                cache.invalidateAll();
+            }
+        }
+    }
+
+    @Override
+    public void invalidate(Partition partition)
+    {
+        if (isCacheEnabledFor(partition.getSchemaTableName()) && isLocationPresent(partition.getStorage())) {
+            cache.invalidate(new Path(partition.getStorage().getLocation()));
+        }
+    }
+
+    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, FileSystem fs, Path path)
+            throws IOException
+    {
+        RemoteIterator<LocatedFileStatus> iterator = fs.listLocatedStatus(path);
```

Comment:
Please move it back to `public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)` method